File: OOP_C++/tema6/lab6_3.cpp
#include <iostream>
#include<string>
using namespace std;

class iLogger {
public:
    virtual void logMessage(string msg) = 0;
    virtual ~iLogger() =default;
};

class hibrid {
    protected:
    string numeDispozitiv;
public:
    string getNume() {
        return numeDispozitiv;
    }

    virtual bool esteConectat() = 0;

    virtual ~hibrid() = default;
};

class Imprimanta:public iLogger, public hibrid {
    private:
    bool conectat;
    public:
    Imprimanta(const string& nume, bool c):conectat(c) {}

    bool esteConectat() override {
        return conectat;
    }

    void logMessage(string msg) override {
        cout<<"[Imprimanta "<<numeDispozitiv<<"]: "<<msg<<endl;
    }


};

int main() {
    Imprimanta imprimanta1("HP LaserJet", true);
    if(imprimanta1.esteConectat()) {
        imprimanta1.logMessage("Imprimanta este conectata si functionala.");
    } else {
        imprimanta1.logMessage("Imprimanta nu este conectata.");
    }

    imprimanta1.logMessage("Incepe imprimarea documentului...");

    return 0;
}

File: OOP_Kotlin/tema6/lab6_3.kt
interface ILogger {
    fun logMessage(msg: String)
}

/*
In tipul de date abstract putem avea si functii concrete
 */

abstract class Dispozitiv(protected val numeDispozitiv: String) {
    fun getNume(): String {
        return numeDispozitiv
    }

    abstract fun esteConectat(): Boolean
}

/*
!!!IMPORTANT
Clasa imprimanta nu face multiple inheritance.Aceasta mosteneste o clasa abstracta si o interfata.
In Kotlin o clasa nu poate mostenii mai multe clase concrete, insa se poate mostenii o clasa si ORICATE interfete.
 */

class Imprimanta(
    nume: String,
    private val conectat: Boolean
) : Dispozitiv(nume), ILogger {

    override fun esteConectat(): Boolean {
        return conectat
    }

    override fun logMessage(msg: String) {
        println("[Imprimanta $numeDispozitiv]: $msg")
    }
}

fun main() {
    val imprimanta1 = Imprimanta("HP LaserJet", true)

    if (imprimanta1.esteConectat()) {
        imprimanta1.logMessage("Imprimanta este conectata si functionala.")
    } else {
        imprimanta1.logMessage("Imprimanta nu este conectata.")
    }

    imprimanta1.logMessage("Incepe imprimarea documentului...")
}
